### PR TITLE
ROX-18448, ROX-18449: Sensor workflow to suspend and resume scan

### PR DESCRIPTION
**Pull request by Vincent056**
## Description

Implementing suspend and resume scan features.

## Checklist
- [ ] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

### Here I tell how I validated my change

TODO(replace-me)  
deploy and test on a cluster

* Added following mock test to sensor:
  * SuspendingScanNotFound: Tested the response when no scan is found for suspending 
  * ResumingScanNotFound:  Tested the response when no scan is found for Resuming 
  * Testing Suspending and Resuming Scan success under normal condition

### Reminder for reviewers
N/A

### CODE REVIEW

File: proto/internalapi/central/compliance_operator.proto [hunk 1]
```diff
@@ -54,12 +54,20 @@ message ApplyComplianceScanConfigRequest {
...
+    message ResumeScheduledScan {
+        string scan_name = 1;
+    }
 
     string                  id              = 1;
     oneof scan_request {
         OneTimeScan         one_time_scan   = 2;
         ScheduledScan       scheduled_scan  = 3;
         RerunScheduledScan  rerun_scan      = 4;
+        SuspendScheduledScan suspend_scan   = 5;
```
Comment (dashrews78) [inline proto/internalapi/central/compliance_operator.proto:69]: nit:  we generally line up the variable names and the numbering.

File: sensor/kubernetes/complianceoperator/handler_impl.go [hunk 2]
```diff
@@ -246,7 +260,86 @@ func (m *handlerImpl) processRerunScheduledScanRequest(requestID string, request
 	}
 	return m.composeAndSendApplyScanConfigResponse(requestID, err)
 }
+func (m *handlerImpl) processScanConfigScheduleChangeRequest(requestID string, config scanScheduleConfiguration) bool {
```
Comment (dashrews78) [inline sensor/kubernetes/complianceoperator/handler_impl.go:263]: Is the idea to use this method when we get around to implementing the update flow as well?  I see schedule is referenced in here, but does not seem to be set or used

Comment (Vincent056) [reply sensor/kubernetes/complianceoperator/handler_impl.go:263]: I think the schedule is updated at https://github.com/stackrox/stackrox/pull/9005/files#diff-934acc4d45abf0e86d3d3656f7dc018e3af42951927a56ddd8815b1d58c6f9a0R303

Comment (Vincent056) [reply sensor/kubernetes/complianceoperator/handler_impl.go:263]: I wonder if there is anything I missed

File: sensor/kubernetes/complianceoperator/utils.go [hunk 3]
```diff
@@ -14,6 +14,19 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 )
 
+type scanNameGetter interface {
+	GetScanName() string
+}
+
+func validateScanName(req scanNameGetter) error {
+	if req == nil {
+		return errors.New("apply scan configuration request is empty")
```
Comment (c-du) [inline sensor/kubernetes/complianceoperator/utils.go:23]: Need to change this. Scan name getter is null?

File: sensor/kubernetes/complianceoperator/handler_impl.go [hunk 4]
```diff
@@ -31,6 +33,14 @@ type handlerImpl struct {
 	stopSignal concurrency.Signal
 }
 
+type scanScheduleConfiguration struct {
+	Suspend        *bool
+	Schedule       *string
+	ScanName       string
+	Request        interface{}
```
Comment (c-du) [inline sensor/kubernetes/complianceoperator/handler_impl.go:40]: Is it too open to have an interface{} here? It should be at least a request.

Comment (Vincent056) [reply sensor/kubernetes/complianceoperator/handler_impl.go:40]: I wonder how would I make it to accept either `ApplyComplianceScanConfigRequest_ResumeScheduledScan` or `ApplyComplianceScanConfigRequest_SuspendScheduledScan`, I do have a validateInterface function to verify the request

File: sensor/kubernetes/complianceoperator/handler_impl.go [hunk 5]
```diff
@@ -241,7 +255,82 @@ func (m *handlerImpl) processRerunScheduledScanRequest(requestID string, request
...
+		err = errors.Wrap(err, "Could not convert unstructured to scan setting")
+		return m.composeAndSendApplyScanConfigResponse(requestID, err)
+	}
+
+	// Check if scanSetting has the "Suspend" field
+	if _, ok := reflect.TypeOf(scanSetting).FieldByName("Suspend"); ok {
+		scanSetting.Suspend = *config.Suspend
+	} else {
+		// Handle the case where the field doesn't exist (older CRD)
+		return m.composeAndSendApplyScanConfigResponse(requestID, errors.New("Compliance operator version is too old"))
```
Comment (dashrews78) [inline sensor/kubernetes/complianceoperator/handler_impl.go:286]:
```suggestion
		return m.composeAndSendApplyScanConfigResponse(requestID, errors.New("suspending a scan is not supported on this version of the compliance operator"))
```

File: sensor/kubernetes/complianceoperator/utils.go [hunk 6]
```diff
@@ -14,6 +14,19 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 )
 
+type scanNameGetter interface {
+	GetScanName() string
+}
+
+func validateScanName(req scanNameGetter) error {
+	if req == nil {
+		return errors.New("scan name getter is nil")
```
Comment (dashrews78) [inline sensor/kubernetes/complianceoperator/utils.go:23]:
```suggestion
		return errors.New("scan name empty")
```

File: sensor/kubernetes/complianceoperator/handler_impl.go [hunk 7]
```diff
@@ -241,7 +255,82 @@ func (m *handlerImpl) processRerunScheduledScanRequest(requestID string, request
...
+		err = errors.Wrap(err, "Could not convert unstructured to scan setting")
+		return m.composeAndSendApplyScanConfigResponse(requestID, err)
+	}
+
+	// Check if scanSetting has the "Suspend" field
+	if _, ok := reflect.TypeOf(scanSetting).FieldByName("Suspend"); ok {
+		scanSetting.Suspend = *config.Suspend
+	} else {
+		// Handle the case where the field doesn't exist (older CRD)
+		return m.composeAndSendApplyScanConfigResponse(requestID, errors.New("suspending a scan is not supported on this version of the compliance operator"))
```
Comment (rhmdnd) [inline sensor/kubernetes/complianceoperator/handler_impl.go:286]: This reminds me - where we going to have a check somewhere that ensured the Compliance Operator was at least version 1.4.0 for the integration to work?

Comment (dashrews78) [reply sensor/kubernetes/complianceoperator/handler_impl.go:286]: https://issues.redhat.com/browse/ROX-18101

It wasn't so much that it was on 1.4.0 but that they were all the same.  Probably worth re-visiting that one in the relatively near future.

**Comment by Vincent056** [comment]
rebase to resolve conflict from master

File: sensor/kubernetes/complianceoperator/handler_impl.go (hunk 4 above)
Comment (c-du) [reply sensor/kubernetes/complianceoperator/handler_impl.go:39]: We agree that this could be done in a separate PR

Comment (Vincent056) [reply sensor/kubernetes/complianceoperator/handler_impl.go:39]: we are going to address this in a separate PR, tracked in https://issues.redhat.com/browse/ROX-21657

**Review by dashrews78 — APPROVED** [review]

**Pull request merged**